libs/datasets/sources/covid_county_data.py

```diff
@@ -40,10 +44,88 @@ class CovidCountyDataDataSource(data_source.DataSource):
+        True            True            True         False  245      Testing data already set
+        False           False           False        False  361      No data available
+        True            False           True         True   627      pos = tests - neg
+        False           True            False        True   1079     Testing data not available
+        True            True            False        True   7064     Pos and neg tests set, total_tests not used
+        False           True            True         True   25769    neg = tests - pos
+        False           False           True         True   32684    pos = cases, neg = tests - pos
+        True            True            True         True   61951    Have everything already
+        False           False           False        True   508238   No test data
+        """
+
+        # Make a copy to avoid modifying the argument when using mask with inplace=True.
+        df = data.copy()
+        tests_and_cases = df.eval(
+            "negative_tests.isna() & positive_tests.isna() & total_tests.notna() & cases.notna()"
+        )
+        missing_neg = df.eval(
+            "negative_tests.isna() & positive_tests.notna() & total_tests.notna() & cases.notna()"
+        )
+        missing_pos = df.eval(
+            "negative_tests.notna() & positive_tests.isna() & total_tests.notna() & cases.notna()"
+        )
+
+        # Keep the same order of rows in `provenance` as `df` so that the same masks can be used when
+        # writing it.
+        provenance = df.loc[:, [CommonFields.FIPS, CommonFields.DATE]]
+        provenance[CommonFields.POSITIVE_TESTS] = "none"
+        provenance[CommonFields.NEGATIVE_TESTS] = "none"
+
+        # Using where/mask is suggested by https://stackoverflow.com/a/36067403/341400
+        df[CommonFields.POSITIVE_TESTS].mask(tests_and_cases, df[CommonFields.CASES], inplace=True)
+        df[CommonFields.NEGATIVE_TESTS].mask(
+            tests_and_cases, df[CommonFields.TOTAL_TESTS] - df[CommonFields.CASES], inplace=True
+        )
+        provenance[CommonFields.POSITIVE_TESTS].mask(
+            tests_and_cases, "tests_and_cases", inplace=True
+        )
+        provenance[CommonFields.NEGATIVE_TESTS].mask(
+            tests_and_cases, "tests_and_cases", inplace=True
+        )
+
+        df[CommonFields.NEGATIVE_TESTS].mask(
+            missing_neg,
+            df[CommonFields.TOTAL_TESTS] - df[CommonFields.POSITIVE_TESTS],
+            inplace=True,
+        )
+        provenance[CommonFields.NEGATIVE_TESTS].mask(missing_neg, "missing_neg", inplace=True)
+
+        df[CommonFields.POSITIVE_TESTS].mask(
+            missing_pos,
+            df[CommonFields.TOTAL_TESTS] - df[CommonFields.NEGATIVE_TESTS],
+            inplace=True,
+        )
+        provenance[CommonFields.POSITIVE_TESTS].mask(missing_pos, "missing_pos", inplace=True)
+
+        provenance_series = libs.datasets.combined_datasets._to_timeseries_rows(
```

Comment:
could you make this function public if you're going to call it from outside of combined_datasets?

Reply:
also, without looking at the code, I'm not totally sure what this does.  What is a timeseries row in this context?

Reply:
made public and renamed